### DISCUSSION TITLE
Export source graph to visualise tech debt / architecture

### DISCUSSION THREAD

**Discussion by jjrscott** [peripheryapp/periphery#758] [Ideas]
For a while I've been running a modified version of periphery (https://github.com/jjrscott/periphery/tree/jjrscott/charts) which exports the SourceGraph to json. I then run this through various scripts and GraphViz to produce charts such as this representation of periphery's Frontend:

![frontend_type_graph](https://github.com/peripheryapp/periphery/assets/528531/f6c6fcec-811b-46af-8f44-2f9a1a9e354f)

My reasoning is to start using these charts to continually reduce overall tech debt in a project by finding and removing unneccessary links (usage) and generally slowly transforming the graph into something "simpler".

I'd like find other interested parties who can advise/assist me in how best to migrate these scripts into the main project. To that end here's a few questions I've been pondering on:

1. Should this function be part of the `scan` command or should I make new `chart` command? There's a lot of similar code and I don't want to duplicate it.
2. Should I stick with GraphViz or switch to Mermaid as Github handles it automatically in READMEs etc
3. Could I use code to make comments on the graph like "type X in target Y may be better placved in a target Z" (definiately feels it should be in `scan`)

Thanks for your time.

**Comment by jjrscott** [comment]
Here's the same chart but in Mermaid:
```mermaid
flowchart LR

s_8Frontend18CheckUpdateCommandV[struct CheckUpdateCommand]
s_8Frontend17ClearCacheCommandV[struct ClearCacheCommand]
s_8Frontend0A7CommandP[protocol FrontendCommand]
s_e_s_8Frontend0A7CommandPAAE11_errorLabelSSvpZ[extension FrontendCommand]
s_8Frontend12ScanBehaviorC[class ScanBehavior]
s_8Frontend11ScanCommandV[struct ScanCommand]
s_e_s_6Shared12OutputFormatOs_14ArgumentParser013ExpressibleByA0P[extension OutputFormat]
s_e_s_13SystemPackage8FilePathV8FrontendE8argumentACSgSS_tcfc[extension FilePath]
s_8Frontend14VersionCommandV[struct VersionCommand]
s_8Frontend16CommonSetupGuideC[class CommonSetupGuide]
s_8Frontend11GuidedSetupC[class GuidedSetup]
s_8Frontend7ProjectC[class Project]
s_8Frontend20SPMProjectSetupGuideC[class SPMProjectSetupGuide]
s_8Frontend4ScanC[class Scan]
s_8Frontend13UpdateCheckerC[class UpdateChecker]
s_8Frontend16PeripheryCommandV[struct PeripheryCommand]
s_8Frontend11GuidedSetupC --> s_8Frontend7ProjectC
s_8Frontend11ScanCommandV --> s_8Frontend0A7CommandP
s_8Frontend11ScanCommandV --> s_8Frontend4ScanC
s_8Frontend12ScanBehaviorC --> s_8Frontend11GuidedSetupC
s_8Frontend12ScanBehaviorC --> s_8Frontend7ProjectC
s_8Frontend14VersionCommandV --> s_8Frontend0A7CommandP
s_8Frontend16PeripheryCommandV --> s_8Frontend0A7CommandP
s_8Frontend16PeripheryCommandV --> s_8Frontend11ScanCommandV
s_8Frontend16PeripheryCommandV --> s_8Frontend14VersionCommandV
s_8Frontend16PeripheryCommandV --> s_8Frontend17ClearCacheCommandV
s_8Frontend16PeripheryCommandV --> s_8Frontend18CheckUpdateCommandV
s_8Frontend17ClearCacheCommandV --> s_8Frontend0A7CommandP
s_8Frontend18CheckUpdateCommandV --> s_8Frontend0A7CommandP
s_8Frontend4ScanC --> s_8Frontend7ProjectC
s_e_s_8Frontend0A7CommandPAAE11_errorLabelSSvpZ --> s_8Frontend0A7CommandP
```

**Comment by Sakababib** [comment]
Hi @jjrscott , is there a way to visualise how many dead lines of code a project has?
